flat_mutation_reader: Split readers by file and remove unnecessary includes.

mikolajsieluzycki opened the pull request:
The flat_mutation_reader files were conflated and contained multiple
readers, which were not strictly necessary. Splitting optimizes both
iterative compilation times, as touching rarely used readers doesn't
recompile large chunks of codebase. Total compilation times are also
improved, as the size of flat_mutation_reader.hh and
flat_mutation_reader_v2.hh have been reduced and those files are
included by many file in the codebase.

With changes - 948 files

real	29m14.051s
user	168m39.071s
sys	5m13.443s

Without changes - 929 files

real	30m36.203s
user	175m43.354s
sys	5m26.376s

denesb commented:
Thanks for doing this @mikolajsieluzycki, this is a long needed cleanup.
While I fully agree with moving each reader into its own header (extra points for the `readers/` dir), I don't agree with splitting each reader into its own `.cc` file. While on the surface this looks like a good thing as it allows for more parallelism while compiling, below a certain size threshold (don't know what that size is exactly) it is counter productive. So we try to keep our `.cc` file few and relatively large and only split when compiling them becomes a pain. You will see that we have many `.cc` files containing the implementation for many different headers. So I think it makes sense to keep all reader implementations in `flat_mutation_reader.cc` and only split the header. Or maybe move all reader implementation into a new `.cc` in `readers/` and leave only strictly `flat_mutation_reader` related stuff in `flat_mutation_reader.cc`.
I think we should similarly split `mutation_reader.hh` (and maybe move the content of `mutation_reader.cc` into the new `.cc` file in `readers/`) but this can be done in a follow-up, this cleanup is good enough on its own.

denesb reviewed (changes requested):
I think moving each reader into its own `.cc` is a mistake, see above.

mikolajsieluzycki commented:
@denesb If we had modules, I'd probably argue for splitting, since the overhead of re-parsing the headers would be minimized. I think moving flat_mutation_reader to one cc and everything else to another would be a reasonable middle ground. I'll update shortly.

mikolajsieluzycki commented:
@denesb Removed multiple cc files and replaced with just 2 - one with flat_mutation_reader[_v2] and the other with everything else.

denesb commented on dht/i_partitioner_fwd.hh:
```diff
@@ -0,0 +1,45 @@
+// the beginning of the token space etc.
+//
+// We'll fold all of that into the token class and push all of the variations
+// into its users.
```
I think this comment can be dropped.

denesb commented on mutation_fragment_fwd.hh:
```diff
@@ -0,0 +1,45 @@
+    // cannot be used.
+    class forwarding_tag;
+    using forwarding = bool_class<forwarding_tag>;
+}
```
I think that if we are moving this, it should be moved to the header which declares the flat mutation reader (`flat_mutation_reader.hh`). It doesn't really belong to `mutation_fragment.hh`.

denesb replied:
Also, orthogonal to this, the comment explaining something should be where the main declaration is, no one will think to look for the documentation in a `_fwd` header.

denesb commented on readers/empty.hh:
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (C) 2017-present ScyllaDB
```
Shouldn't new files get the current year? (I don't really know)

denesb commented on readers/empty.hh:
```diff
@@ -0,0 +1,18 @@
+class reader_permit;
+
+flat_mutation_reader make_empty_flat_reader(schema_ptr s, reader_permit permit);
+flat_mutation_reader_v2 make_empty_flat_reader_v2(schema_ptr s, reader_permit permit);
```
This is also present in `empty_v2.hh`.

denesb reviewed (approved):
Minor comments.
Opinion: maybe we should replace `readers/` with `mutation_reader/` to mirror the existing `mutation_writer/` folder.
LGTM overall, thanks again for this long-needed but arduous cleanup.

mikolajsieluzycki commented:
Pushed new version:
- changed all new files copyright notice to current year (except renamed files)
- removed make_empty_flat_reader_v2 from readers/empty.hh and updated associated files
- removed comments as suggested from _fwd files
- moved streaming_mutation to flat_mutation_reader.hh
- simplified mutation_reader::forwarding by removing unnecessary and barely used partition_range_forwarding declaration from flat_mutation_reader[_v2] and moving it to mutation_reader namespace and naming it "forwarding" so that we don't have 2 aliases for the same thing.
